### PR TITLE
[MODEL] convert gpt2 model

### PR DESCRIPTION
**Pull request by hymzoque**
## Description ##
convert gpt2 model

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: src/gluonnlp/models/gpt2.py [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+    cfg = gpt2_124M()
+    cfg.defrost()
+    cfg.MODEL.num_heads = 20
+    cfg.MODEL.num_layers = 36
+    cfg.MODEL.units = 1280
+    cfg.freeze()
+    return cfg
+
+PRETRAINED_URL = {
+    'gpt2_124M': {
```
Comment (ZheyuYe) [inline src/gluonnlp/models/gpt2.py:100]: We may call it `openai_gpt_124M`?

**Review by ZheyuYe — COMMENTED** [review]
Would please also update the README

File: src/gluonnlp/models/gpt2.py (hunk 1 above)
Comment (sxjscience) [reply src/gluonnlp/models/gpt2.py:100]: I think GPT2 sounds good

File: src/gluonnlp/models/gpt2.py [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+                   weight_initializer=weight_initializer,
+                   bias_initializer=bias_initializer,
+                   dtype=dtype,
+                   output_all_encodings=output_all_encodings,
+                   layout=cfg.MODEL.layout,
+                   compute_layout=cfg.MODEL.compute_layout)
+
+@use_np
+class GPT2ForLM(HybridBlock):
+    def __init__(self, backbone_cfg=gpt2_124M()):
```
Comment (sxjscience) [inline src/gluonnlp/models/gpt2.py:552]: You may set the default to None

**Comment by sxjscience** [comment]
Add some test cases in test_models_gpt2.py

**Comment by sxjscience** [comment]
One test you may consider is to check that these two cases will generate the same states:

- feed the input tokens one by one
- feed K tokens at a time

File: scripts/conversion_toolkits/convert_gpt2.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+python3 -m pip install tensorflow==1.15 --upgrade --user
+git clone https://github.com/openai/gpt-2.git gpt_2
+for model in 124M 355M 774M
+do
+    python3 gpt_2/download_model.py ${model}
+    mkdir gpt2_${model}
+    CUDA_VISIBLE_DEVICES="" python3 convert_gpt2.py --tf_model_path models/${model} --save_dir gpt2_${model} --test
+done
```
Comment (sxjscience) [inline scripts/conversion_toolkits/convert_gpt2.sh:8]: Add this file to convert_all.sh and update the reader.

**Comment by hymzoque** [comment]
> One test you may consider is to check that these two cases will generate the same states:
> 
> * feed the input tokens one by one
> * feed K tokens at a time

solved

File: tests/test_models_gpt2.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    cfg = GPT2Model.get_cfg()
+    gpt2_model = GPT2Model.from_cfg(cfg)
+    gpt2_model.initialize(ctx=ctx)
+    gpt2_model.hybridize()
+
+    _, one_time_states = gpt2_model(
+        inputs,
+        gpt2_model.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
```
Comment (sxjscience) [inline tests/test_models_gpt2.py:96]: Also test the logits output.

File: tests/test_models_gpt2.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        )
+    incremental_states = states
+    incremental_hiddens = hiddens
+    assert_allclose(incremental_states.asnumpy(),
+                    states.asnumpy(), 1E-4, 1E-4)
+    assert_allclose(incremental_hiddens.asnumpy(),
+                    hiddens.asnumpy(), 1E-4, 1E-4)
+
+
+#@pytest.mark.remote_required
```
Comment (sxjscience) [inline tests/test_models_gpt2.py:113]: No need to comment.

File: tests/test_models_gpt2.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+@pytest.mark.parametrize('model_name', list_pretrained_gpt2())
+def test_gpt2(model_name):
```
Comment (sxjscience) [inline tests/test_models_gpt2.py:115]: You may refer to test_models_bert.py to add the ctx support. You can add a ctx flag to in the arguments like https://github.com/dmlc/gluon-nlp/blob/66e5e057347c3710eb8fa27c134a9303309e94c0/tests/test_models_bert.py#L94.

So you can test the model on both cpu and GPU.
